sw/device/silicon_creator/lib/drivers/kmac.c

```diff
@@ -0,0 +1,279 @@
+  // KMAC should never move out of the `absorb` state until `CMD.PROCESS` is
+  // issued, so we get significant performance gains by polling only once here
+  // instead of before every `absorb`.
+  HARDENED_RETURN_IF_ERROR(poll_state(KMAC_STATUS_SHA3_ABSORB_BIT));
+
+  return kErrorOk;
+}
+
+inline void kmac_shake256_absorb(const uint8_t *in, size_t inlen) {
+  // This implementation does not poll `STATUS.fifo_depth` as recommended in
+  // the KMAC documentation. Normally, polling is required to prevent a
+  // deadlock scenario between Ibex, KMAC, and EDN. However, in this case it is
+  // safe to skip because `kmac_shake256_configure()` sets KMAC to use
+  // software-only entropy, and sets `kmac_en` to false (so KMAC will not
+  // produce entropy requests anyway). Since KMAC will therefore not block on
+  // EDN, it is guaranteed to keep processing message blocks. For more details,
+  // see the KMAC documentation:
+  //   https://docs.opentitan.org/hw/ip/kmac/doc/#fifo-depth-and-empty-status
+
+  // Use byte-wide writes until the input pointer is aligned.
+  // Note: writes to the KMAC message FIFO are not required to be aligned.
+  ptrdiff_t offset = misalignment32_of((uintptr_t)in);
+  size_t idx = 0;
+  if (offset != 0) {
+    size_t nbytes = alignof(uint32_t) - offset;
+    for (; idx < nbytes && idx < inlen; idx++) {
+      abs_mmio_write8(kBase + KMAC_MSG_FIFO_REG_OFFSET, in[idx]);
+    }
+  }
+
+  // Use word writes for all full words.
+  for (; idx + sizeof(uint32_t) <= inlen; idx += sizeof(uint32_t)) {
+    abs_mmio_write32(kBase + KMAC_MSG_FIFO_REG_OFFSET, read_32(&in[idx]));
+  }
+
+  // Use byte-wide writes for anything left over.
+  for (; idx < inlen; idx++) {
+    abs_mmio_write8(kBase + KMAC_MSG_FIFO_REG_OFFSET, in[idx]);
+  }
+  HARDENED_CHECK_EQ(idx, inlen);
+}
+
+inline void kmac_shake256_absorb_words(const uint32_t *in, size_t inlen) {
+  // This implementation does not poll `STATUS.fifo_depth` as recommended in
+  // the KMAC documentation. Normally, polling is required to prevent a
+  // deadlock scenario between Ibex, KMAC, and EDN. However, in this case it is
+  // safe to skip because `kmac_shake256_configure()` sets KMAC to use
+  // software-only entropy, and sets `kmac_en` to false (so KMAC will not
+  // produce entropy requests anyway). Since KMAC will therefore not block on
+  // EDN, it is guaranteed to keep processing message blocks. For more details,
+  // see the KMAC documentation:
+  //   https://docs.opentitan.org/hw/ip/kmac/doc/#fifo-depth-and-empty-status
+
+  size_t idx = 0;
+  for (idx = 0; idx < inlen; idx++) {
+    abs_mmio_write32(kBase + KMAC_MSG_FIFO_REG_OFFSET, in[idx]);
+  }
+  HARDENED_CHECK_EQ(idx, inlen);
+}
+
+inline void kmac_shake256_squeeze_start(void) {
+  // Issue `CMD.PROCESS` to move to the squeezing state.
+  issue_command(KMAC_CMD_CMD_VALUE_PROCESS);
+}
+
+rom_error_t kmac_shake256_squeeze_end(uint32_t *out, size_t outlen) {
```

Comment:
please consider my comments above for this function as well.

Reply:
For this one I have an additional hardening check on `idx` that I had trouble replicating when I removed the index. It also seems like there's less benefit for this function since it operates on full words only, so I left it as-is for now. Let me know if you'd like me to revisit.